File: lib/codingChallenges.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import logger from './logger'

export const SNIPPET_PATHS = Object.freeze(['./server.ts', './routes', './lib', './data', './data/static/web3-snippets', './frontend/src/app', './models', './infrastructure'])

interface FileMatch {
  path: string
  content: string
}

interface CachedCodeChallenge {
  snippet: string
  vulnLines: number[]
  neutralLines: number[]
}

export const findFilesWithCodeChallenges = async (paths: readonly string[]): Promise<FileMatch[]> => {
  const matches = []
  for (const currPath of paths) {
    try {
      if ((await fs.lstat(currPath)).isDirectory()) {
        const files = await fs.readdir(currPath)
        const moreMatches = await findFilesWithCodeChallenges(
          files.map(file => path.resolve(currPath, file))
        )
        matches.push(...moreMatches)
      } else {
        const code = await fs.readFile(currPath, 'utf8')
        if (
          // strings are split so that it doesn't find itself...
          code.includes('// vuln-code' + '-snippet start') ||
          code.includes('# vuln-code' + '-snippet start')
        ) {
          matches.push({ path: currPath, content: code })
        }
      }
    } catch (e) {
      logger.warn(`File ${currPath} could not be read. It might have been moved or deleted. If coding challenges are contained in the file, they will not be available.`)
    }
  }

  return matches
}

function getCodeChallengesFromFile (file: FileMatch) {
  const fileContent = file.content

  // get all challenges which are in the file by a regex capture group
  const challengeKeyRegex = /[#\/]{0,2} vuln-code-snippet start (?<challenges>[^\r\n]*)/g
  const challenges = [...fileContent.matchAll(challengeKeyRegex)]
    .flatMap(match => match.groups?.challenges?.trim().split(/\s+/) ?? [])
    .filter(Boolean)

  return challenges.map((challengeKey) => getCodingChallengeFromFileContent(fileContent, challengeKey))
}

export function getCodingChallengeFromFileContent (source: string, challengeKey: string) {
  const startMarker = `vuln-code-snippet start ${challengeKey}`
  const endMarker = `vuln-code-snippet end ${challengeKey}`
  const startIdx = source.indexOf(startMarker)
  const endIdx = source.indexOf(endMarker, startIdx === -1 ? 0 : startIdx)
  if (startIdx === -1 || endIdx === -1) {
    throw new BrokenBoundary('Broken code snippet boundaries for: ' + challengeKey)
  }
  let snippet = source.slice(startIdx, endIdx + endMarker.length) // TODO Currently only a single code snippet is supported
  let firstNewline = snippet.indexOf('\r\n')
  let nlLen = 2
  if (firstNewline === -1) {
    firstNewline = snippet.indexOf('\n')
    nlLen = 1
  }
  if (firstNewline === -1) {
    firstNewline = snippet.indexOf('\r')
    nlLen = 1
  }
  if (firstNewline !== -1) snippet = snippet.slice(firstNewline + nlLen)
  const endMarkerIdx = snippet.lastIndexOf(endMarker)
  if (endMarkerIdx !== -1) {
    snippet = snippet.slice(0, endMarkerIdx)
    if (snippet.endsWith('\r\n')) snippet = snippet.slice(0, -2)
    else if (snippet.endsWith('\n') || snippet.endsWith('\r')) snippet = snippet.slice(0, -1)
  }
  let normalized = snippet.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
  const hideStart = 'vuln-code-snippet hide-start'
  const hideEnd = 'vuln-code-snippet hide-end'
  let idx = normalized.indexOf(hideStart)
  while (idx !== -1) {
    const endHide = normalized.indexOf(hideEnd, idx)
    if (endHide === -1) {
      normalized = normalized.slice(0, idx)
      break
    }
    normalized = normalized.slice(0, idx) + normalized.slice(endHide + hideEnd.length)
    idx = normalized.indexOf(hideStart)
  }
  let lines = normalized.split('\n').filter(line => !line.includes('vuln-code-snippet hide-line'))
  const vulnLines = []
  const neutralLines = []
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    if (line.indexOf('vuln-code-snippet vuln-line') !== -1 && line.indexOf(challengeKey) !== -1) {
      vulnLines.push(i + 1)
    } else if (line.indexOf('vuln-code-snippet neutral-line') !== -1 && line.indexOf(challengeKey) !== -1) {
      neutralLines.push(i + 1)
    }
  }
  lines = lines.filter(line => !(line.indexOf('vuln-code-snippet vuln-line') !== -1 && line.indexOf(challengeKey) !== -1) && !(line.indexOf('vuln-code-snippet neutral-line') !== -1 && line.indexOf(challengeKey) !== -1))
  snippet = lines.join('\n').trim()
  return { challengeKey, snippet, vulnLines, neutralLines }
}

export class BrokenBoundary extends Error {
  constructor (message: string) {
    super(message)
    this.name = 'BrokenBoundary'
    this.message = message
  }
}

// dont use directly, use getCodeChallenges getter
let _internalCodeChallenges: Map<string, CachedCodeChallenge> | null = null
export async function getCodeChallenges (): Promise<Map<string, CachedCodeChallenge>> {
  if (_internalCodeChallenges === null) {
    _internalCodeChallenges = new Map<string, CachedCodeChallenge>()
    const filesWithCodeChallenges = await findFilesWithCodeChallenges(SNIPPET_PATHS)
    for (const fileMatch of filesWithCodeChallenges) {
      for (const codeChallenge of getCodeChallengesFromFile(fileMatch)) {
        _internalCodeChallenges.set(codeChallenge.challengeKey, {
          snippet: codeChallenge.snippet,
          vulnLines: codeChallenge.vulnLines,
          neutralLines: codeChallenge.neutralLines
        })
      }
    }
  }
  return _internalCodeChallenges
}

File: lib/coding_challenges.py
import os
import re
import logging
import pandas as pd

logger = logging.getLogger(__name__)

SNIPPET_PATHS = [
    './server.py',
    './routes',
    './lib',
    './data',
    './data/static/web3-snippets',
    './frontend/src/app',
    './models'
]

class BrokenBoundary(Exception):
    def __init__(self, message):
        super().__init__(message)
        self.name = 'BrokenBoundary'
        self.message = message

def find_files_with_code_challenges(paths):
    matches = []
    for curr_path in paths:
        if os.path.isdir(curr_path):
            files = [os.path.join(curr_path, file) for file in os.listdir(curr_path)]
            more_matches = find_files_with_code_challenges(files)
            matches.extend(more_matches)
        else:
            try:
                with open(curr_path, 'r', encoding='utf-8') as file:
                    code = file.read()
                    if (
                        '// vuln-code-snippet start' in code or
                        '# vuln-code-snippet start' in code
                    ):
                        matches.append({'path': curr_path, 'content': code})
            except Exception as e:
                logger.warn(f"File {curr_path} could not be read. it might have been moved or deleted. If coding challenges are contained in the file, they will not be available.")
    return matches

def get_code_challenges_from_file(file):
    file_content = file['content']
    challenge_key_regex = re.compile(r'[/#]{0,2} vuln-code-snippet start (?P<challenges>.*)')
    challenges = [
        match.group('challenges').split(' ')
        for match in challenge_key_regex.finditer(file_content)
    ]
    challenges = [item for sublist in challenges for item in sublist if item]
    return [get_coding_challenge_from_file_content(file_content, challenge_key) for challenge_key in challenges]

def get_coding_challenge_from_file_content(source, challenge_key):
    snippets = re.findall(rf'[/#]{{0,2}} vuln-code-snippet start.*{challenge_key}([^])*vuln-code-snippet end.*{challenge_key}', source)
    if not snippets:
        raise BrokenBoundary(f'Broken code snippet boundaries for: {challenge_key}')
    snippet = snippets[0]
    snippet = re.sub(r'\s?[/#]{0,2} vuln-code-snippet start.*[\r\n]{0,2}', '', snippet)
    snippet = re.sub(r'\s?[/#]{0,2} vuln-code-snippet end.*', '', snippet)
    snippet = re.sub(r'.*[/#]{0,2} vuln-code-snippet hide-line[\r\n]{0,2}', '', snippet)
    snippet = re.sub(r'.*[/#]{0,2} vuln-code-snippet hide-start([^])*[/#]{0,2} vuln-code-snippet hide-end[\r\n]{0,2}', '', snippet)
    snippet = snippet.strip()

    lines = snippet.split('\r\n')
    if len(lines) == 1:
        lines = snippet.split('\n')
    if len(lines) == 1:
        lines = snippet.split('\r')
    vuln_lines = []
    neutral_lines = []
    for i, line in enumerate(lines):
        if re.search(rf'vuln-code-snippet vuln-line.*{challenge_key}', line):
            vuln_lines.append(i + 1)
        elif re.search(rf'vuln-code-snippet neutral-line.*{challenge_key}', line):
            neutral_lines.append(i + 1)
    snippet = re.sub(r'\s?[/#]{0,2} vuln-code-snippet vuln-line.*', '', snippet)
    snippet = re.sub(r'\s?[/#]{0,2} vuln-code-snippet neutral-line.*', '', snippet)
    return {'challenge_key': challenge_key, 'snippet': snippet, 'vuln_lines': vuln_lines, 'neutral_lines': neutral_lines}

_internal_code_challenges = None

def get_code_challenges():
    global _internal_code_challenges
    if _internal_code_challenges is None:
        _internal_code_challenges = {}
        files_with_code_challenges = find_files_with_code_challenges(SNIPPET_PATHS)
        for file_match in files_with_code_challenges:
            for code_challenge in get_code_challenges_from_file(file_match):
                _internal_code_challenges[code_challenge['challenge_key']] = {
                    'snippet': code_challenge['snippet'],
                    'vuln_lines': code_challenge['vuln_lines'],
                    'neutral_lines': code_challenge['neutral_lines']
                }
    return _internal_code_challenges
